### PR TITLE
DRILL-2288: Fix ScanBatch violation of IterOutcome protocol and downstream chain of bugs

### PR DESCRIPTION
**Pull request by dsbos**

### CODE REVIEW

File: contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseRecordReader.java [hunk 1]
```diff
@@ -142,10 +148,12 @@ public void setup(OperatorContext context, OutputMutator output) throws Executio
           getOrCreateFamilyVector(column.getRootSegment().getPath(), false);
         }
       }
-      logger.debug("Opening scanner for HBase table '{}', Zookeeper quorum '{}', port '{}', znode '{}'.",
-          hbaseTableName, hbaseConf.get(HConstants.ZOOKEEPER_QUORUM),
-          hbaseConf.get(HBASE_ZOOKEEPER_PORT), hbaseConf.get(HConstants.ZOOKEEPER_ZNODE_PARENT));
-      hTable = new HTable(hbaseConf, hbaseTableName);
+      // Add vector for any column families not mentioned yet (in order to avoid
+      // creation of dummy NullableIntVectors for them).
+      for (HColumnDescriptor columnFamily :
```
Comment (jacques-n) [inline contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseRecordReader.java]: This also needs to be done for any requested column vectors.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java [hunk 2]
```diff
@@ -325,10 +325,13 @@ public AggOutcome doWork() {
                 if (EXTRA_DEBUG_1) {
                   logger.debug("Received new schema.  Batch has {} records.", incoming.getRecordCount());
                 }
-                newSchema = true;
-                this.cleanup();
-                // TODO: new schema case needs to be handled appropriately
-                return AggOutcome.UPDATE_AGGREGATOR;
+                final BatchSchema newIncomingSchema = incoming.getSchema();
+                if ((! newIncomingSchema.equals(schema)) && schema != null) {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java]: You are quelling a new schema here without correctly reloading vector references. A schema change means: the schema has changed and/or the vectors have changed. Your check here isn't enough to guarantee correct behavior. It is possible that the schema is equal but the vectors changed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java [hunk 3]
```diff
@@ -122,6 +122,7 @@ protected void killIncoming(final boolean sendUpstream) {
 
   @Override
   public IterOutcome innerNext() {
+    recordCount = 0;
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java]: Why don't you add this to parent class rather than each child of the parent? (Either AbstractRecordBatch or AbstractSingleRecordBatch.)

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java [hunk 4]
```diff
@@ -335,7 +343,8 @@ public IterOutcome nextBatch() throws SchemaChangeException {
             return iterRight;
 
           default:
-            throw new IllegalStateException(String.format("Unknown state %s.", iterRight));
+            throw new IllegalStateException(
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Why did you change the formatting here?

File: exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatch.java [hunk 5]
```diff
@@ -23,60 +23,214 @@
 import org.apache.drill.exec.record.selection.SelectionVector4;
 
 /**
- * A record batch contains a set of field values for a particular range of records. In the case of a record batch
- * composed of ValueVectors, ideally a batch fits within L2 cache (~256k per core). The set of value vectors do not
- * change unless the next() IterOutcome is a *_NEW_SCHEMA type.
- *
- * A key thing to know is that the Iterator provided by record batch must align with the rank positions of the field ids
- * provided utilizing getValueVectorId();
+ * A record batch contains a set of field values for a particular range of
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatch.java]: Nice documentation on this code.

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/AbstractMapVector.java [hunk 6]
```diff
@@ -189,8 +189,8 @@ protected void putVector(String name, ValueVector vector) {
         Preconditions.checkNotNull(vector, "vector cannot be null")
     );
     if (old != null && old != vector) {
-      logger.debug("Field [%s] mutated from [%s] to [%s]", name, old.getClass().getSimpleName(),
-          vector.getClass().getSimpleName());
+      logger.debug("Field [{}] mutated from [{}] to [{}]", name, old.getClass().getSimpleName(),
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/AbstractMapVector.java]: I think there is a limit to the number of bracket replacements on the logger interface. Is three allowed?

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/MapVector.java [hunk 7]
```diff
@@ -309,7 +309,14 @@ public Object getObject(int index) {
       Map<String, Object> vv = new JsonStringHashMap<>();
       for (String child:getChildFieldNames()) {
         ValueVector v = getChild(child);
-        if (v != null) {
+        // TODO(DRILL-4001):  Resolve this hack:
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/MapVector.java]: Can you add what happens if this hack isn't done? What is the error?

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java [hunk 8]
```diff
@@ -93,13 +93,15 @@ public void ensureAtLeastOneField(ComplexWriter writer) {
...
+      PathSegment fieldPath = sp.getRootSegment();
       BaseWriter.MapWriter fieldWriter = writer.rootAsMap();
-      while (root.getChild() != null && !root.getChild().isArray()) {
-        fieldWriter = fieldWriter.map(root.getNameSegment().getPath());
-        root = root.getChild();
+      while (fieldPath.getChild() != null && ! fieldPath.getChild().isArray()) {
+        fieldWriter = fieldWriter.map(fieldPath.getNameSegment().getPath());
+        fieldPath = fieldPath.getChild();
+      }
+      if (fieldWriter.isEmptyMap()) {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: can you explain this change?

Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: That should be just a renaming:

It looked like the variable originally named "root" related to the root (from getRootSegment()) only until it was modified in the loop (so the name was misleading from the beginning of the loop and especially at the last reference).

Renaming it with "field" was to reflect that it corresponds to variable "fieldWriter" even as both are possibly modified in the loop.

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/AbstractMapVector.java (hunk 6 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/AbstractMapVector.java]: http://www.slf4j.org/api/org/slf4j/helpers/MessageFormatter.htmlDoesn't the number of brace replacement patterns correspond to the number of arguments?

File: exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatch.java (hunk 5 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatch.java]: Thanks.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java (hunk 4 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Hmm.  It's been a while, so I don't recall exactly.  It looks like I wrapped it to fit within 80 columns since most of the code and comments just above fit within that width.  (Yes, now I see that the next throw statement below is also wider.)

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java (hunk 3 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java]: That wouldn't have worked--the recordCount fields are declared in the subclasses, not in one of those superclasses. 

(I didn't try refactoring recordCount up into one of those base classes because I was already running into enough downstream problems every time I fixed an upstream increment.)

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java (hunk 2 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java]: Yeah, I thought this change might not be right.

So what do I do (to at least avoid the schema-change exception that I was getting from one of the test suites, but do so without introducing the bug you point out)?

File: contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseRecordReader.java (hunk 1 above)
Comment (dsbos) [inline contrib/storage-hbase/src/main/java/org/apache/drill/exec/store/hbase/HBaseRecordReader.java]: Created DRILL-4010 and added unit test (currently ignored) for it.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java (hunk 2 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java]: You shouldn't put a hack in to ignore a valid schema change. I don't know all the details but if this is related to HBase, my thought is you probably need to resolve 4010 to avoid spurious schema changes. You should make sure to avoid propagating schema changes that are not real/required. Solving on a particular operator just means if the query plan changes, you'll probably get an error in a different operator.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java (hunk 3 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java]: Got it. Fair enough.

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java (hunk 8 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: The change I was focused on was the addition of the fieldWriter.isEmptyMap() condition. Why was this condition added to the code?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java (hunk 2 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java]: > You shouldn't put a hack in to ignore a valid schema change.

Yeah, I'm reverting this HashAggTemplate change.

>  I don't know all the details but if this is related to HBase, my thought is you probably need
>  to resolve 4010 to avoid spurious schema changes. 

It's not clear whether it's related to HBase or not.  (The relevant test failure did occur in an HBase test, but it was in a Jenkins-run cluster test that has been hard to reproduce elsewhere.  At one point (I thought) I had a reproduction using just JSON files, but now that doesn't seem to work.)

> You should make sure to avoid propagating schema changes that are not real/required. 

Related to that, ExternalSortBatch seems to return two OK_NEW_SCHEMA batches with schemas differing only in their selection vectors: 

11:23:50.149 [main] TRACE o.a.d.e.p.i.v.IteratorValidatorBatchIterator - [#3; on ExternalSortBatch]: incoming next() return: #records = 204, 
  schema:
    BatchSchema [fields=[`key1`(INT:REQUIRED), `alt`(BIGINT:REQUIRED)], selectionVector=FOUR_BYTE], 
  prev. new (not equal):
    BatchSchema [fields=[`key1`(INT:REQUIRED), `alt`(BIGINT:REQUIRED)], selectionVector=NONE]

(The earlier schema ("prev. new:") seems to come from setting up the sort, and the latter one ("schema:") seems to come from completing the sort (it was triggered by a NONE from upstream).)

Is that correct?  (Should the selection vectors have been the same?  Should the second return have been OK instead of OK_NEW_SCHEMA?) 

If that is correct (e.g., vectors were changed and should have been), how should downstream operators, or at least  aggregation, handle that?

Should they be able to handle a "technical" schema change (OK_NEW_SCHEMA with possibly-changed vectors or certain details of the schema) that is not a logical schema change (not a change in the logical fields or types in the aggregated value), even if arbitrary logical schema changes can't be handled (e.g., because some don't even make sense)?

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/MapVector.java (hunk 7 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/MapVector.java]: Added mention of MapVector.getObject(int) and exception message "java.lang.IndexOutOfBoundsException: index: 0, length: 1 (expected: range(0, 0))" to DRILL-4001 report.

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java (hunk 8 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: Recall that ensureAtLeastOneField used to create that dummy NullableIntVector field/column (so that the schema didn't end up with zero fields/columns) _regardless_ of whether it needed to (i.e., even if the schema already had a column or columns from a previous JsonRecordReader for the same ScanBatch).  

Under the failure conditions, that blind vector setting frequently replaced a NullableVarCharVector with that NullableIntVector, causing either explicit "schema change not supported" user-targeted errors or internal errors about the mismatch between the NullableIntVector and the expectation of a NullableVarCharVector.

The additional check prevents ensureAtLeastOneField from overriding a column in an existing non-empty schema and causing those schema changes.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java (hunk 2 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggTemplate.java]: Reverted via a13e99b ("2288: Pt. 11 Core Upd.: REVERTED all of bad OK_NEW_SCHEMA schema comp…")

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java (hunk 8 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: If you make this change, isn't this sufficient to remove entirely the ensureAtLeastOneField operations and simply check this condition?

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/AbstractMapVector.java (hunk 6 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/AbstractMapVector.java]: ok

**Comment by dsbos** [comment]
Note:  Something went wrong between me and Git.  I intended to address review comments only by adding commits (e.g., rather than by, say, deleting any), in order to avoid obscuring earlier comments here in the GitHub conversation.  However, for some reason, more than just the last four update commits (with "Upd.") show up as being added today.  I apparently re-ordered some commits during rebasing.

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java (hunk 8 above)
Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: Sorry; I can't yet tell what you mean.

(By "the ensureAtLeastOneField operations," did you mean operations _in_ ensureAtLeastOneField (e.g., calling map(...) (in the loop)), the operation of calling ensureAtLeastOneField, or something else? (The first seems to be needed to get to the right map, the second seems to be needed to create a dummy column when one is actually needed.))

Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: Let me restate:

We're using ensureAtLeastOneField to make sure we have at least one vector has been created. If we check this condition and it is false, then we do nothing. My question is: can we remove the ensureAtLeastOneField property and replace the condition above to be fieldWriter.isEmptyMap() instead?

Basically change:

if (!atLeastOneWrite) {

to

if (fieldWriter.isEmptyMap()) {

Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: > can we remove the ensureAtLeastOneField property  

Oh, you mean the atLeastOneWrite property?

Yes, I think that will work.  I'll test that..

Comment (dsbos) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: Another question:

> Basically change:
> 
> if (!atLeastOneWrite) {
> 
> to
> 
> if (fieldWriter.isEmptyMap()) {

Would that field writer be the writer from writer.rootAsMap(), or does the code still need the loop and its traversal?

(It seems that changing the "if (!atLeastOneWrite)" to (the simplified equivalent) of "if (true)" (and keeping the loop) work work, but I don't know what the minimum fix is (i.e., exactly how much of that loop and traversal is needed).)

**Comment by dsbos** [comment]
Rebased on today's master branch.

File: exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java (hunk 8 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java]: @hnfgns , thoughts on this?

**Comment by dsbos** [comment]
Addressed problems revealed by today's rebasing.
